File: frontend/src/content/skill.ts
// Per-environment skill thresholds: map a run's Score → a skill band.
// CartPole's Score is mean episode length (max 500). As of E2 the live thresholds come from
// the backend (`GET /api/skill/{env}` → `scaleFromEnvSkill`, single source of truth); the table
// below is only a fallback used until that fetch lands (and a default for unknown envs).

import type { EnvSkill, EnvSpec, SkillBandId } from '../api/types'

export type SkillKey = 'child' | 'below' | 'average' | 'above' | 'superhuman'

// Backend band ids (schemas/skill.py) → the shorter frontend keys the i18n + UI already use.
const BAND_ID_TO_KEY: Record<SkillBandId, SkillKey> = {
  child:         'child',
  below_average: 'below',
  average:       'average',
  above_average: 'above',
  superhuman:    'superhuman',
}

export interface SkillBand {
  key: SkillKey
  /** inclusive lower bound on Score for this band */
  min: number
}

export interface SkillScale {
  /** bottom of the bar (0% fill) — 0 for CartPole, negative for shaped envs like LunarLander */
  min: number
  max: number
  bands: SkillBand[] // ascending by min; first band's min == scale.min
}

export const SKILL_SCALES: Record<string, SkillScale> = {
  cartpole: {
    min: 0,
    max: 500,
    bands: [
      { key: 'child', min: 0 },
      { key: 'below', min: 50 },
      { key: 'average', min: 150 },
      { key: 'above', min: 300 },
      { key: 'superhuman', min: 450 },
    ],
  },
}

export const DEFAULT_SKILL_SCALE: SkillScale = SKILL_SCALES.cartpole

export function skillScaleFor(envId: string | null): SkillScale {
  return (envId !== null && SKILL_SCALES[envId]) || DEFAULT_SKILL_SCALE
}

// Band lower bounds as a fraction of the env's [min_score, solved_score] range — must match the
// backend's _BAND_FRACTIONS (services/skill.py).
const BAND_FRACTIONS: { key: SkillKey; frac: number }[] = [
  { key: 'child', frac: 0 },
  { key: 'below', frac: 0.1 },
  { key: 'average', frac: 0.3 },
  { key: 'above', frac: 0.6 },
  { key: 'superhuman', frac: 0.95 },
]

/** Build the skill scale from an env's own registry spec (solved_score + min_score). The robust
 *  fallback when the backend thresholds haven't loaded yet — so the meter always uses the SELECTED
 *  env's range and never another env's (e.g. CartPole's 0–500) leaks onto LunarLander. */
export function scaleFromEnvSpec(env: EnvSpec | undefined): SkillScale | null {
  if (!env) return null
  const min = env.min_score ?? 0
  const max = env.solved_score
  const span = max - min
  return { min, max, bands: BAND_FRACTIONS.map((b) => ({ key: b.key, min: min + b.frac * span })) }
}

/** Widen a scale's 0% floor for longer **play** episodes (EnvSpec.play_step_scale). A
 *  step-penalty env played at 3× the steps has a ~3× deeper failure floor, while a *success*
 *  score is unchanged — so multiply min by the factor and recompute the bands over the new range
 *  (must mirror the backend's `env_skill(min_scale=...)`). Only meaningful for min < 0; a no-op at
 *  factor 1. */
export function scaleForPlay(scale: SkillScale, factor: number): SkillScale {
  if (factor === 1 || scale.min >= 0) return scale
  const min = scale.min * factor
  const span = scale.max - min
  return { min, max: scale.max, bands: BAND_FRACTIONS.map((b) => ({ key: b.key, min: min + b.frac * span })) }
}

/** Convert the backend's per-env thresholds into the local SkillScale shape (the meter's
 *  bands + the rating now share one source: the env's solved_score). */
export function scaleFromEnvSkill(envSkill: EnvSkill): SkillScale {
  return {
    min: envSkill.min_score ?? 0,  // default guards an older backend that predates min_score
    max: envSkill.max_score,
    bands: envSkill.bands.map((b) => ({ key: BAND_ID_TO_KEY[b.id], min: b.min_score })),
  }
}

/** The highest band whose threshold the score has reached. */
export function currentBand(score: number, scale: SkillScale): SkillBand {
  let band = scale.bands[0]
  for (const b of scale.bands) if (score >= b.min) band = b
  return band
}

File: backend/app/services/skill.py
"""Skill evaluation — turn a play-session score into a beginner-friendly band.

Thresholds are **data-driven from the env registry**: each band's lower bound is a fixed
fraction of the env's ``solved_score`` (100% of the goal). So CartPole (solved=500) yields
child<50, below-avg<150, avg<300, above-avg<475, superhuman≥475, and any future game with a
``solved_score`` gets sensibly scaled bands for free — no per-env skill config to maintain.

The fraction scheme assumes a non-negative score that climbs toward ``solved_score`` (true of
CartPole and the Atari/Box2D envs on the desktop roadmap). A symmetric-score game like Pong
(−21…21) would want a custom band table; when that lands, add an optional ``skill_bands``
override to :class:`~app.envs.registry.EnvSpec` and prefer it here. Until then this stays
deliberately small and registry-sourced.
"""

from app.envs.registry import get_env
from app.schemas.skill import EnvSkill, SkillBand, SkillBandId, SkillRating

# Band lower bounds as a fraction of the env's solved_score, weakest → strongest. Documented
# here (and surfaced verbatim via GET /api/skill/{env}) so the thresholds are explicit.
_BAND_FRACTIONS: list[tuple[SkillBandId, float]] = [
    ("child", 0.00),
    ("below_average", 0.10),
    ("average", 0.30),
    ("above_average", 0.60),
    ("superhuman", 0.95),
]


def env_skill(env_id: str, min_scale: float = 1.0) -> EnvSkill | None:
    """The concrete band thresholds for ``env_id``, or ``None`` if the env is unknown.

    Bands span the env's ``[min_score, solved_score]`` range, so a shaped env that starts in
    the red (LunarLander, ``min_score=-100``) gets bands that climb through the negatives
    instead of all bunching at 0.

    ``min_scale`` widens the floor for longer **play** episodes: an env played at
    ``play_step_scale=3`` runs 3× the steps, so its failure floor (≈ −1 × max_steps) is ~3×
    deeper while a *success* score is unchanged — pass the play scale so the meter/rating span
    matches the longer episode (default 1.0 = the standard training span).
    """
    spec = get_env(env_id)
    if spec is None:
        return None
    max_score = spec.solved_score
    min_score = spec.min_score * min_scale
    span = max_score - min_score
    bands = [
        SkillBand(id=band_id, min_score=round(min_score + frac * span, 4))
        for band_id, frac in _BAND_FRACTIONS
    ]
    return EnvSkill(env_id=env_id, max_score=max_score, min_score=min_score, bands=bands)


def rate(env_id: str, score: float, min_scale: float = 1.0) -> SkillRating | None:
    """Rate a finished session's ``score`` against ``env_id``'s bands.

    Returns the highest band whose ``min_score`` the score reaches (the lowest band starts at
    the env's ``min_score``, so a score is always rated), plus a clamped fill ratio for the
    meter measured across ``[min_score, solved_score]``. ``None`` only if the env is unknown.
    ``min_scale`` widens the floor for longer play episodes (see :func:`env_skill`).
    """
    skill = env_skill(env_id, min_scale)
    if skill is None:
        return None
    band: SkillBandId = skill.bands[0].id
    for candidate in skill.bands:
        if score >= candidate.min_score:
            band = candidate.id
    span = skill.max_score - skill.min_score
    ratio = (score - skill.min_score) / span if span > 0 else 0.0
    ratio = max(0.0, min(1.0, ratio))
    return SkillRating(band=band, score=score, max_score=skill.max_score, ratio=ratio)
